### DISCUSSION TITLE
Export not working

### DISCUSSION THREAD

**Discussion by fhsdmhacc** [filamentphp/filament#13372] [Help]
### Package

Actions

### Package Version

v3.2

### How can we help you?

I have an exporter, but when the export is finished, I want to download it to .xlsx format and redirect to another page and it shows "this site can't be reached", but when i download it to .csv format, the .csv downloaded and if i open the file it shows blank sheets. please help

![Screenshot 2024-06-24 120107](https://github.com/filamentphp/filament/assets/103301588/c9ee02ce-0549-4af0-b1f7-e42fd8ee6245)

**Comment by leandrocfe** [comment]
Please share your code

**Reply by fhsdmhacc** [reply]
![image](https://github.com/filamentphp/filament/assets/103301588/5f91d208-eb30-4475-a7ba-517e9caa74c2)

**Reply by sh4msi** [reply]
Did you find a solution?

**Reply by fhsdmhacc** [reply]
no, i use another plugin https://github.com/pxlrbt/filament-excel

**Comment by christopherlllobrera** [comment]
Do you have Laravel excel  package installed?

**Reply by fhsdmhacc** [reply]
not yet.
does it need to be installed first?

**Comment by fhsdmhacc** [comment]
the error from log

production.ERROR: fpassthru(): Argument #1 ($stream) must be of type resource, null given

**Comment by sh4msi** [comment]
I still get this error on the server in version 3.2.96
But it works locally without any problems

Is there a solution?

 @danharrin  @leandrocfe

**Reply by Alhiane** [reply]
> I still get this error on the server in version 3.2.96 But it works locally without any problems
> 
> Is there a solution?
> 
> @danharrin @leandrocfe

Same it works localy for both, csv and xslx
in production dosn't works, it's only download empty files
and when i check via consol i can see that files are there and they have correct data, but when i click on download from my panel
i got empty file

**Comment by juse-less** [comment]
I stumbled upon this error, as well.
CSV files being downloaded empty, and XLSX files causing a server error.

In my case, it was because of file permission issues.
I had to change the values for Laravel's public disk, per the documentation.
https://laravel.com/docs/11.x/filesystem#local-files-and-visibility

Although, if a different, local, disk is used, then that one needs the configuration instead.

Exactly which permissions are needed depends on setup.
If the web server/PHP owns the files, then something like the default in the documentation should w.
```php
'permissions' => [
    'file' => [
        'public' => 0644,
        'private' => 0600,
    ],
    'dir' => [
        'public' => 0755,
        'private' => 0700,
    ],
]
```

If the web server/PHP group owns the files, then you'd need to change the private part to something like
```php
'permissions' => [
    'file' => [
        'public' => 0644,
        'private' => 0640,
    ],
    'dir' => [
        'public' => 0755,
        'private' => 0750,
    ],
]
```

And so on.
I haven't quite figured out why this is, yet, though.
It writes the exports perfectly fine to disk for me, but can't read it when downloading.
In my case, the web server and PHP is even the same user and group (httpd running PHP through mod_php).

**Comment by ttebify** [comment]
In addition to Juse-Less answer above, creating a new disk specifically for exporting resolved the issue for me. Here’s how you can configure it:

in `config/filesystem.php`

```php
'export' => [ // Custom disk for exports
    'driver' => 'local',
    'root' => storage_path('app/exports'),
    'url' => env('APP_URL') . '/exports',
    'visibility' => 'public',
    'permissions' => [
        'file' => [
            'public' => 0644,
            'private' => 0600,
        ],
        'dir' => [
            'public' => 0755,
            'private' => 0700,
        ],
    ],
    'throw' => false,
],
```

Make sure to add the `FILAMENT_FILESYSTEM_DISK=export` to your environment variables as mentioned in the documentation:

https://filamentphp.com/docs/3.x/actions/prebuilt-actions/export#customizing-the-storage-disk

**Reply by raphael-talenavi** [reply]
i already done this but its still not working

**Reply by ttebify** [reply]
> i already done this but its still not working

Well, as you search for solutions, keep in mind that it's a permission issue.

Check the ownership of the disk and confirm that your web server has access to it. I eventually opted to use an AWS driver in production, and the issue never arose.

However, when using a local disk, permission issues may arise, as newly created files by Filament are owned by root. If that solution doesn't work, I'm not sure how else I can assist.

**Reply by raphael-talenavi** [reply]
in my case it was working if the private files is set to 0644 and the private dir to 0755, i dont know why this happen but the visibility is public

**Comment by hinimajesty** [comment]
I chased down this issue in the past but didn't find a fix. I think i have finally found what was wrong in my case. It turns out that for those of us using queues monitored by supervisor, each queue worker is started by the user configured within the supervisor program. So assuming the user in the program is root the exported file will be created and owned by root. Nginx however has its own user called www-data. By clicking on the download link of the file www-data (nginx) is trying to access a file owned by root which wont work. The solution is to have same www-data user within your supervisor program. This way the file created can be accessed by nginx as well. Things to try . 

Navigate to `{your app directory}/storage/apps/exports/filament_exports/{the number in the broken file link}`

Then ls -l {fileName} this will give you the owner of the file. Then run ps -aux and check the user of nginx. If they dont match then clearly the queue worker that exported the file is owned by another user. 

Example: 
Assuming my app resides in /var/www/myapp.com and the broken download link to my file is `myapp.com/filament/exports/9` , i will cd into `/var/www/myapp.com/storage/apps/exports/filament_exports/9` and `ls -l filename`. This command will reveal the owner of the file. Additionally i will cd into /etc/supervisor/conf.d and nano worker.conf or horizon.conf etc depending on how you setup supervisor to monitor your queue workers. In that file i will check to ensure the user of the exported file and the user of the queue worker are same. 

My complete setup (I am not sure all these will be required if you fix the supervisor program to have same user as nginx)

1. I setup filesystems.php to include a disk called export purposely for filament. 

 ```
'export' => [ 
            'driver' => 'local',
            'root' => storage_path('app/exports'),
            'url' => env('APP_URL') . '/exports',
            'visibility' => 'public',
            'permissions' => [
                'file' => [
                    'public' => 0644,
                    'private' => 0600,
                ],
                'dir' => [
                    'public' => 0755,
                    'private' => 0700,
                ],
            ],
            'throw' => false,
        ]
```
2. I added FILAMENT_FILESYSTEM_DISK=export to .env
3. My supervisor program 

```
[program:horizon]
process_name=%(program_name)s
command=php /var/www/myapp.com/artisan horizon
autostart=true
autorestart=true
user=www-data
redirect_stderr=true
stdout_logfile=/var/www/myapp.com/storage/logs/horizon.log
stopwaitsecs=3600
```

**Reply by urvish022** [reply]
Thank you @hinimajesty, Your solution works for me, that was horizon user/ownership problem.

**Reply by johnymanuelli** [reply]
@hinimajesty Thanks, changing the supervisor user fixed my problem.

**Reply by OccTherapist** [reply]
Thank you very much. Changing the supervisor user was the fix for my app, too.
